managed/CounterStrikeSharp.API/Core/GlobalContext.cs

```diff
@@ -19,27 +19,71 @@
 using System.IO;
 using System.Linq;
 using System.Reflection;
+using System.Runtime.Loader;
 using System.Text;
 using CounterStrikeSharp.API.Modules.Admin;
 using CounterStrikeSharp.API.Modules.Commands;
 using CounterStrikeSharp.API.Modules.Menu;
 using CounterStrikeSharp.API.Modules.Utils;
+using McMaster.NETCore.Plugins;
 
 namespace CounterStrikeSharp.API.Core
 {
     public sealed class GlobalContext
     {
-        private static GlobalContext _instance = null;
-        public static GlobalContext Instance => _instance;
+        public static GlobalContext Instance { get; private set; }
 
-        private DirectoryInfo rootDir;
+        private readonly string _rootDir;
 
         private readonly List<PluginContext> _loadedPlugins = new();
 
+        public PluginApiRegistry ApiRegistry { get; } = new();
+
+        private readonly List<(AssemblyName name, Assembly assembly)> _sharedAssemblies = new();
+
+        private readonly AssemblyLoadContext _pluginLoadContext = new("plugin_load_context");
+
         public GlobalContext()
         {
-            rootDir = new FileInfo(Assembly.GetExecutingAssembly().Location).Directory.Parent;
-            _instance = this;
+            _rootDir = new FileInfo(Assembly.GetExecutingAssembly().Location).Directory!.Parent!.FullName;
+            Instance = this;
+
+            var defaultContext = AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly())!;
+
+            _pluginLoadContext.Resolving += (context, name) =>
+            {
+                try
+                {
+                    return defaultContext.LoadFromAssemblyName(name);
+                }
+                catch
+                {
+                    var assembly = _sharedAssemblies.First(a => a.name.Name == name.Name);
+                    return assembly.assembly;
+                }
+            };
+
+            LoadSharedLibs();
+        }
```

Comment:
`McMaster` has its own implementation of shared libraries so I'd say this is needless

Reply:
> `McMaster` has its own implementation of shared libraries so I'd say this is needless

what exactly are you talking about?

Reply:
> > `McMaster` has its own implementation of shared libraries so I'd say this is needless
> 
> what exactly are you talking about?

https://www.nuget.org/packages/McMaster.NETCore.Plugins

They use this package to load plugin assemblies.